openedx/core/djangoapps/content/search/engine.py

```diff
@@ -0,0 +1,492 @@
+
+You will then need to create the new indices by running:
+
+    ./manage.py lms shell -c "from openedx.core.djangoapps.content.search import engine; engine.create_indexes()"
+
+For more information about the Meilisearch API in Python, check
+https://github.com/meilisearch/meilisearch-python
+
+When implementing a new index, you might discover that you need to list explicit filterable
+fields. Typically, you try to index new documents, and Meilisearch fails with the
+following response:
+
+    meilisearch.errors.MeilisearchApiError: MeilisearchApiError. Error code: invalid_search_filter.
+    Error message: Attribute `field3` is not filterable. Available filterable attributes are:
+    `field1 field2 _pk`.
+
+In such cases, the filterable field should be added to INDEX_FILTERABLES below. And you should
+then run the `create_indexes()` function again, as indicated above.
+
+This search engine was tested for the following indexes:
+
+1. course_info ("course discovery"):
+    - Enable the course discovery feature: FEATURES["ENABLE_COURSE_DISCOVERY"] = True
+    - A search bar appears in the LMS landing page.
+    - Content is automatically indexed every time a course is edited in the studio.
+2. courseware_content ("courseware search"):
+    - Enable the courseware search waffle flag:
+
+        ./manage.py lms waffle_flag --create --everyone courseware.mfe_courseware_search
+
+    - Enable the following feature flags:
+
+        FEATURES["ENABLE_COURSEWARE_INDEX"] = True
+        FEATURES["ENABLE_COURSEWARE_SEARCH"] = True
+    - Reindex courseware content by running: ./manage.py cms reindex_course --active
+    - Alternatively, reindex course content by clicking the "Reindex" button in the Studio.
+    - In the learning MFE, a course search bar appears when opening a course.
+"""
+
+from copy import deepcopy
+from datetime import datetime
+import hashlib
+import json
+import logging
+import typing as t
+
+import meilisearch
+
+from django.conf import settings
+from django.utils import timezone
+
+from search.search_engine_base import SearchEngine
+from search.utils import ValueRange
+
+
+logger = logging.getLogger(__name__)
+
+PRIMARY_KEY_FIELD_NAME = "_pk"
+UTC_OFFSET_SUFFIX = "__utcoffset"
+
+
+# In Meilisearch, we need to explicitly list fields for which we expect to define
+# filters and aggregation functions.
+# This is different than Elasticsearch where we can aggregate results over any field.
+# Here, we list facet fields per index.
+# Reference: https://www.meilisearch.com/docs/learn/filtering_and_sorting/search_with_facet_filters
+# Note that index names are hard-coded here, because they are hardcoded anyway across all of edx-search.
+INDEX_FILTERABLES = {
+    "course_info": [
+        "language",  # aggregate by language, mode, org
+        "modes",
+        "org",
+        "catalog_visibility",  # exclude visibility="none"
+        "enrollment_end",  # include only enrollable courses
+    ],
+    "courseware_content": [
```

Comment:
Should use [CoursewareSearchIndexer.INDEX_NAME](https://github.com/openedx/edx-platform/blob/e28a01e2bf78d5750c649edf5e8d48776cc38d8d/cms/djangoapps/contentstore/courseware_index.py#L343) instead of a hard-coded string here.

There's surely a variable for `"course_info"` too, but I couldn't find it quickly.

It's unfortunate that we need to store the filterable fields for all the uses of this engine here, instead of near the code that uses them, but I don't see a better way to deal with this..

Should we provide a way for people to inject their own index + fields into this hash?